### PR TITLE
tcproute/udproute support multiple backends

### PR DESCRIPTION
**Pull request by zirain**
fix: https://github.com/envoyproxy/gateway/issues/3174

### CODE REVIEW

File: internal/gatewayapi/route.go [hunk 1]
```diff
@@ -846,24 +846,22 @@ func (t *Translator) processUDPRouteParentRefs(udpRoute *UDPRouteContext, resour
...
+			// Skip further processing if route destination is not valid
+			if ds == nil || len(ds.Endpoints) == 0 {
+				valid = false
+				continue
+			}
+
+			destSettings = append(destSettings, ds)
+		}
+		if !valid {
 			continue
```
Comment (arkodg) [inline internal/gatewayapi/route.go:862]: we should add a status for the `invalid` case

Comment (zirain) [reply internal/gatewayapi/route.go:862]: correct me if I'm wrong, looks like we didn't do that thing for all routes?

File: internal/gatewayapi/route.go [hunk 2]
```diff
@@ -846,24 +846,22 @@ func (t *Translator) processUDPRouteParentRefs(udpRoute *UDPRouteContext, resour
...
-		}
 
-		backendRef := udpRoute.Spec.Rules[0].BackendRefs[0]
-		ds, _ := t.processDestination(backendRef, parentRef, udpRoute, resources)
-		// Skip further processing if route destination is not valid
-		if ds == nil || len(ds.Endpoints) == 0 {
+		valid := true
+		for _, backendRef := range udpRoute.Spec.Rules[0].BackendRefs {
+			ds, _ := t.processDestination(backendRef, parentRef, udpRoute, resources)
+			// Skip further processing if route destination is not valid
```
Comment (arkodg) [inline internal/gatewayapi/route.go:853]: imo we should follow the same approach we use in HTTP case, where for an invalid destination, we track the weight as invalid, and build a weighted cluster based on valid and invalid weights across the multiple backendRefs

File: internal/gatewayapi/route.go [hunk 3]
```diff
@@ -978,23 +975,21 @@ func (t *Translator) processTCPRouteParentRefs(tcpRoute *TCPRouteContext, resour
...
-				"One and only one backend is supported",
-			)
-			continue
-		}
 
-		backendRef := tcpRoute.Spec.Rules[0].BackendRefs[0]
-		ds, _ := t.processDestination(backendRef, parentRef, tcpRoute, resources)
-		// Skip further processing if route destination is not valid
-		if ds == nil || len(ds.Endpoints) == 0 {
+		valid := true
```
Comment (arkodg) [inline internal/gatewayapi/route.go:979]: same as above

File: internal/gatewayapi/route.go (hunk 2 above)
Comment (zirain) [reply internal/gatewayapi/route.go:853]: May I know what's the benifit for doing that?

Comment (arkodg) [reply internal/gatewayapi/route.go:853]: its defined in the spec https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1alpha2.TCPRouteRule

Comment (zirain) [reply internal/gatewayapi/route.go:853]: seems different to https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPBackendRef?

Comment (arkodg) [reply internal/gatewayapi/route.go:853]: also mentioned in https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPRouteRule

File: internal/xds/translator/route.go [hunk 4]
```diff
@@ -221,27 +218,30 @@ func buildXdsStringMatcher(irMatch *ir.StringMatch) *matcherv3.StringMatcher {
...
 }
 
-func buildXdsRouteAction(httpRoute *ir.HTTPRoute) *routev3.RouteAction {
-	return &routev3.RouteAction{
-		ClusterSpecifier: &routev3.RouteAction_Cluster{
-			Cluster: httpRoute.Destination.Name,
-		},
-		IdleTimeout: idleTimeout(httpRoute),
+func buildXdsRouteAction(backendWeights *ir.BackendWeights) *routev3.RouteAction {
+	if backendWeights.Invalid == 0 {
```
Comment (arkodg) [inline internal/xds/translator/route.go:222]: can we add a comment here saying we only use a weighted cluster when we have invalid weights

File: internal/xds/translator/route.go [hunk 5]
```diff
@@ -71,13 +71,10 @@ func buildXdsRoute(httpRoute *ir.HTTPRoute) (*routev3.Route, error) {
...
-		var routeAction *routev3.RouteAction
-		if httpRoute.BackendWeights.Invalid != 0 {
-			// If there are invalid backends then a weighted cluster is required for the route
-			routeAction = buildXdsWeightedRouteAction(httpRoute)
-		} else {
-			routeAction = buildXdsRouteAction(httpRoute)
-		}
+		backendWeights := httpRoute.Destination.ToWeightedBackend()
+		routeAction := buildXdsRouteAction(backendWeights)
+		routeAction.IdleTimeout = idleTimeout(httpRoute)
```
[the remark this answers is not in the document]
Comment (arkodg) [reply internal/xds/translator/route.go:76]: thanks for adding this note :). the diff is massive, stared with this file and the IR and this approach makes sense

File: internal/ir/xds.go [hunk 6]
```diff
@@ -926,14 +925,33 @@ func (r RouteDestination) Validate() error {
 	return errs
 }
 
+func (r *RouteDestination) ToWeightedBackend() *BackendWeights {
```
Comment (arkodg) [inline internal/ir/xds.go:928]: why not `ToBackendWeights` ?

File: internal/gatewayapi/testdata/custom-filter-order.out.yaml [hunk 7]
```diff
@@ -236,11 +236,10 @@ xdsIR:
         mergeSlashes: true
       port: 10080
       routes:
-      - backendWeights:
-          invalid: 1
-          valid: 0
-        directResponse:
```
Comment (arkodg) [inline internal/gatewayapi/testdata/custom-filter-order.out.yaml:242]: why is `directResponse` getting removed ?

File: internal/gatewayapi/testdata/backendtrafficpolicy-status-conditions.out.yaml [hunk 8]
```diff
@@ -575,11 +574,10 @@ xdsIR:
         mergeSlashes: true
       port: 10080
       routes:
-      - backendWeights:
-          invalid: 1
-          valid: 0
-        directResponse:
```
Comment (arkodg) [inline internal/gatewayapi/testdata/backendtrafficpolicy-status-conditions.out.yaml:581]: why is `directResponse` getting removed

File: internal/gatewayapi/testdata/gateway-with-listener-with-tcproute-with-mismatch-port-protocol.out.yaml [hunk 9]
```diff
@@ -94,3 +94,10 @@ xdsIR:
     - address: 0.0.0.0
       name: envoy-gateway/gateway-1/tcp
       port: 10162
+      routes:
+      - destination:
+          name: tcproute/default/tcproute-1/rule/-1
+          settings:
+          - weight: 1
+        name: tcproute/default/tcproute-1
+        tls: {}
```
Comment (arkodg) [inline internal/gatewayapi/testdata/gateway-with-listener-with-tcproute-with-mismatch-port-protocol.out.yaml:103]: curious why we have an empty `tls` struct here

File: internal/gatewayapi/testdata/gateway-with-listener-with-udproute-with-multiple-backends.in.yaml [hunk 10]
```diff
@@ -26,8 +26,8 @@ udpRoutes:
       rules:
         - backendRefs:
             - name: service-1
-              port: 8080
+              port: 8162
```
Comment (arkodg) [inline internal/gatewayapi/testdata/gateway-with-listener-with-udproute-with-multiple-backends.in.yaml:29]: how did the port change here ?

File: internal/ir/xds.go [hunk 11]
```diff
@@ -926,14 +925,33 @@ func (r RouteDestination) Validate() error {
...
+	}
+
+	return w
+}
+
 // DestinationSetting holds the settings associated with the destination
 // +kubebuilder:object:generate=true
 type DestinationSetting struct {
 	// Weight associated with this destination.
-	// Note: Weight is not used in TCP/UDP route.
```
Comment (arkodg) [inline internal/ir/xds.go:933]: can we add a comment here saying that invalid endpoints are represents with a non zero weight with an empty endpoints list ?

File: internal/gatewayapi/testdata/backendtrafficpolicy-status-conditions.out.yaml (hunk 8 above)
Comment (zirain) [reply internal/gatewayapi/testdata/backendtrafficpolicy-status-conditions.out.yaml:579]: fixed, see https://github.com/envoyproxy/gateway/pull/3212/commits/115fdb1f3692ff97846f523c4706fba25eb59ab8

File: internal/gatewayapi/route.go [hunk 12]
```diff
@@ -516,7 +516,8 @@ func (t *Translator) processGRPCRouteRules(grpcRoute *GRPCRouteContext, parentRe
 
 		// If the route has no valid backends then just use a direct response and don't fuss with weighted responses
 		for _, ruleRoute := range ruleRoutes {
-			if ruleRoute.Destination == nil && ruleRoute.Redirect == nil {
+			noValidBackends := ruleRoute.Destination == nil || ruleRoute.Destination.ToBackendWeights().Valid == 0
```
Comment (arkodg) [inline internal/gatewayapi/route.go:519]: shouldnt this be `invalidBackendsExist` instead of `noValidBackends`  i.e. some instead of all ?

Comment (zirain) [reply internal/gatewayapi/route.go:519]: noValidBackends = invalidBackendsExist | emptyBackends

File: internal/gatewayapi/testdata/gateway-with-listener-with-tcproute-with-mismatch-port-protocol.out.yaml (hunk 9 above)
Comment (zirain) [reply internal/gatewayapi/testdata/gateway-with-listener-with-tcproute-with-mismatch-port-protocol.out.yaml]: see https://github.com/envoyproxy/gateway/pull/3212/commits/9084108b4c3ac8736a0a17a9764f534029e71261

File: internal/gatewayapi/route.go (hunk 12 above)
Comment (arkodg) [reply internal/gatewayapi/route.go:519]: but `noValidBackends` also implies validBackends == 0 which is not true here

File: internal/gatewayapi/route.go [hunk 13]
```diff
@@ -1158,12 +1136,12 @@ func (t *Translator) processDestination(backendRefContext BackendRefContext,
 
 	backendNamespace := NamespaceDerefOr(backendRef.Namespace, route.GetNamespace())
 	if !t.validateBackendRef(backendRefContext, parentRef, route, resources, backendNamespace, routeType) {
-		return nil, weight
+		return &ir.DestinationSetting{Weight: &weight}
```
Comment (arkodg) [inline internal/gatewayapi/route.go:1139]: can we add a comment here highlighting that this is referring to an invalid weight

File: internal/gatewayapi/route.go [hunk 14]
```diff
@@ -493,30 +493,31 @@ func (t *Translator) processGRPCRouteRules(grpcRoute *GRPCRouteContext, parentRe
...
 				}
+				route.Destination.Settings = append(route.Destination.Settings, ds)
 			}
 		}
 
 		// If the route has no valid backends then just use a direct response and don't fuss with weighted responses
 		for _, ruleRoute := range ruleRoutes {
-			if ruleRoute.Destination == nil && ruleRoute.Redirect == nil {
+			noValidBackends := ruleRoute.Destination == nil || ruleRoute.Destination.ToBackendWeights().Valid == 0
+			if noValidBackends && ruleRoute.Redirect == nil {
```
Comment (arkodg) [inline internal/gatewayapi/route.go:520]: isnt something like this needed in `processHTTPRouteRules` as well ?

File: internal/gatewayapi/route.go (hunk 12 above)
Comment (zirain) [reply internal/gatewayapi/route.go:519]: in current code validBackends == 0 equals emptyBackends

File: internal/gatewayapi/route.go (hunk 14 above)
Comment (zirain) [reply internal/gatewayapi/route.go:520]: good catch

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by shawnh2 — APPROVED** [review]
lgtm

**Pull request merged**